src/backend/columnar/columnar_customscan.c

```diff
@@ -655,18 +1160,77 @@ ColumnarScan_CreateCustomScanState(CustomScan *cscan)
 	CustomScanState *cscanstate = &columnarScanState->custom_scanstate;
 	cscanstate->methods = &ColumnarScanExecuteMethods;
 
-	if (EnableColumnarQualPushdown)
+	return (Node *) cscanstate;
+}
+
+
+static Node *
+EvalParamsMutator(Node *node, ExprContext *econtext)
+{
+	if (node == NULL)
 	{
-		columnarScanState->qual = cscan->scan.plan.qual;
+		return NULL;
 	}
 
-	return (Node *) cscanstate;
+	if (IsA(node, Param))
+	{
+		Param *param = (Param *) node;
+		int16 typLen;
+		bool typByVal;
+		bool isnull;
+
+		get_typlenbyval(param->paramtype, &typLen, &typByVal);
+
+		/* XXX: should save ExprState for efficiency */
+		ExprState *exprState = ExecInitExprWithParams((Expr *) node,
+													  econtext->ecxt_param_list_info);
+		Datum pval = ExecEvalExpr(exprState, econtext, &isnull);
+
+		return (Node *) makeConst(param->paramtype,
+								  param->paramtypmod,
+								  param->paramcollid,
+								  (int) typLen,
+								  pval,
+								  isnull,
+								  typByVal);
+	}
+
+	return expression_tree_mutator(node, EvalParamsMutator, (void *) econtext);
+}
+
+
+/*
+ * EvaluateClauseParams simply replaces Params with Consts in clauses using
+ * econtext.
+ */
+static List *
+EvaluateClauseParams(ExprContext *econtext, List *clauses)
+{
+	return (List *) expression_tree_mutator(
```

Comment:
Shouldn't we just call `EvalParamsMutator((Node *) clauses, econtext)` here ? EvalParamsMutator anyway calls `expression_tree_mutator`.

Reply:
Done.